File: Data Structures & Algorithms/surrounded-regions/submission-1.java
class Solution {

    private int ROWS;
    private int COLS;

    public void solve(char[][] board) {

        ROWS = board.length;
        COLS = board[0].length;

        for ( int r = 0; r < ROWS; r++ ) {
            if (board[r][0] == 'O') {
                dfs(board, r ,0);
            }
            if (board[r][COLS - 1] == 'O') {
                dfs(board, r ,COLS - 1);
            }
        }

        for (int c = 0; c < COLS; c++) {
            if (board[0][c] == 'O') {
                dfs(board, 0, c);
            }
            if (board[ROWS - 1][c] == 'O') {
                dfs(board, ROWS - 1, c);
            }
        }

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (board[r][c] == 'O') {
                    board[r][c] = 'X';
                } else if (board[r][c] == 'T') {
                    board[r][c] = 'O';
                }
            }
        }     
    }

    private void dfs(char[][] board, int r, int c) {

        // Base Case
        if ( r < 0 || c < 0 || r >= ROWS || c >= COLS || board[r][c] != 'O') {
            return;
        }

        // Recursive case
        board[r][c] = 'T';
        dfs(board,r+1,c);
        dfs(board,r-1,c);
        dfs(board,r,c+1);
        dfs(board,r,c-1);
    }
}

File: Data Structures & Algorithms/valid-tree/submission-1.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        
        if len(edges) > n - 1:
            return False

        adj = [[] for i in range(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        visited = set()
        def dfs(node, par):

            # Base Case
            if node in visited:
                return False

            # Recursive Case
            
            visited.add(node)
            # For all neighbours of current node
            for nei in adj[node]:
                # If we reach current node from parrent, then skip
                if nei == par:
                    continue
                if not dfs(nei, node):
                    return False
            return True

        return dfs(0, -1) and n == len(visited)




